src/hotspot/share/code/nmethod.cpp

```diff
@@ -1262,20 +1316,21 @@ nmethod::nmethod(
     _deopt_mh_handler_offset = 0;
     _unwind_handler_offset   = 0;
 
-    _metadata_offset         = checked_cast<uint16_t>(align_up(code_buffer->total_oop_size(), oopSize));
-    _dependencies_offset     = checked_cast<uint16_t>(_metadata_offset + align_up(code_buffer->total_metadata_size(), wordSize));
-    _scopes_pcs_offset       = _dependencies_offset;
+    CHECKED_CAST(_metadata_offset, uint16_t, (align_up(code_buffer->total_oop_size(), oopSize)));
+    DEBUG_ONLY( int data_end_offset = _metadata_offset + align_up(code_buffer->total_metadata_size(), wordSize); )
+    assert((data_offset() + data_end_offset) <= nmethod_size, "wrong nmethod's size: %d < %d", nmethod_size, (data_offset() + data_end_offset));
+
+    // native wrapper does not have read-only data but we need unique not null address
+    _immutable_data          = data_end();
+    _immutable_data_size     = 0;
+    _nul_chk_table_offset    = 0;
+    _handler_table_offset    = _nul_chk_table_offset;
+    _scopes_pcs_offset       = _handler_table_offset;
     _scopes_data_offset      = _scopes_pcs_offset;
-    _handler_table_offset    = _scopes_data_offset;
-    _nul_chk_table_offset    = _handler_table_offset;
 #if INCLUDE_JVMCI
-    _speculations_offset     = _nul_chk_table_offset;
+    _speculations_offset     = _scopes_data_offset;
     _jvmci_data_offset       = _speculations_offset;
```

Comment:
Why not use 0 for all these?

Reply:
Right. Before all these offsets were assigned to _dependencies_offset which was not 0.
But now we can use 0 for all of them since "_dependencies_offset" is 0.